NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/mcxmap.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "mcxmap" 1 "16 May 2014" "mcxmap 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcxmap \- permute or remap the indices of graphs and matrices\&.
.SH SYNOPSIS

\fBmcxmap\fP
\fB-imx\fP fname (\fIinput\fP)
\fB[-o\fP fname (\fIoutput\fP)\fB]\fP
\fB[-make-map\fP (\fIoutput map file name\fP)\fB]\fP
\fB[-make-mapc\fP (\fIoutput map file name\fP)\fB]\fP
\fB[-make-mapr\fP (\fIoutput map file name\fP)\fB]\fP
\fB[-cmul\fP a (\fIcoefficient\fP)\fB]\fP
\fB[-cshift\fP b (\fItranslate\fP)\fB]\fP
\fB[-rmul\fP c (\fIcoefficient\fP)\fB]\fP
\fB[-rshift\fP d (\fItranslate\fP)\fB]\fP
\fB[-mul\fP e (\fIcoefficient\fP)\fB]\fP
\fB[-shift\fP f (\fItranslate\fP)\fB]\fP
\fB[-map\fP fname (\fIrow/col map file\fP)\fB]\fP
\fB[-rmap\fP fname (\fIrow map file\fP)\fB]\fP
\fB[-cmap\fP fname (\fIcolumn map file\fP)\fB]\fP
\fB[-mapi\fP fname (\fIrow/col map file (use inverse)\fP)\fB]\fP
\fB[-rmapi\fP fname (\fIrow map file (use inverse)\fP)\fB]\fP
\fB[-cmapi\fP fname (\fIcolumn map file (use inverse)\fP)\fB]\fP
\fB[-tab\fP fname (\fIread (and map) tab file\fP)\fB]\fP
.SH DESCRIPTION

This utility relabels graphs or matrices\&. Its main use is in applying a
\fImap file\fP to a given matrix or graph\&. A map file contains a so called
\fImap matrix\fP in mcl format that has some special properties (given
further below)\&. The functionality of mcxmap can also be provided by \fBmcx\fP, as
a mapped matrix (i\&.e\&. the result of applying a map matrix to another matrix)
is simply the usual matrix product of a matrix and a map matrix\&. However,
\fBmcx\fP will construct a new matrix and leave the original matrix to be mapped
alone\&. When dealing with huge matrices, considerable gains in efficiency
memory-wise and time-wise can be achieved by doing the mapping in-place\&.
This is what mcxmap does\&. In the future, its functionality may be
embedded in \fBmcx\fP with new mcx operators\&.

The special properties of a map matrix are

.ZJ 1m 1m "\(bu"
The column domain and row domain are
of the same cardinality\&.
.in -2m
.ZJ 1m 1m "\(bu"
Each column has exactly one entry\&.
.in -2m
.ZJ 1m 1m "\(bu"
Each row domain index occurs in exactly one column\&.
.in -2m

These properties imply that the matrix can be used
as a map from the column domain onto the row domain\&.
An example map matrix is found in the \fBEXAMPLES\fP Section\&.
.SH OPTIONS

.ZI 2m "\fB-o\fP fname (\fIoutput file\fP)"
\&
.br
Output file\&.
.in -2m

.ZI 2m "\fB-imx\fP fname (\fIinput file\fP)"
\&
.br
Input file\&.
.in -2m

.ZI 2m "\fB-map\fP fname (\fIrow/col map file)\fP)"
\&
'in -2m
.ZI 2m "\fB-rmap\fP fname (\fIrow map file\fP)"
\&
'in -2m
.ZI 2m "\fB-cmap\fP fname (\fIcolumn map file\fP)"
\&
'in -2m
.ZI 2m "\fB-mapi\fP fname (\fIrow/col map file (use inverse)\fP)"
\&
'in -2m
.ZI 2m "\fB-rmapi\fP fname (\fIrow map fil (use inverse)\fP)"
\&
'in -2m
.ZI 2m "\fB-cmapi\fP fname (\fIcolumn map fil (use inverse)\fP)"
\&
'in -2m
'in +2m
\&
.br
Different ways to specify map files\&.
.in -2m

.ZI 2m "\fB-make-map\fP (\fIoutput map file name\fP)"
\&
'in -2m
.ZI 2m "\fB-make-mapc\fP (\fIoutput map file name\fP)"
\&
'in -2m
.ZI 2m "\fB-make-mapr\fP (\fIoutput map file name\fP)"
\&
'in -2m
'in +2m
\&
.br
Generate a map that maps the specified domain onto
the appropriate canonical domain and write the map
matrix to file\&.
.in -2m

.ZI 2m "\fB-cmul\fP a (\fIcoefficient\fP)"
\&
'in -2m
.ZI 2m "\fB-cshift\fP b (\fItranslate\fP)"
\&
'in -2m
'in +2m
\&
.br
These options have affect if neither a column map file nor column
canonification is specified\&. If any of the first two options is used,
column indices\ \&\fCi\fP are mapped to\ \&\fCa*i+b\fP\&.
.in -2m

.ZI 2m "\fB-rmul\fP c (\fIcoefficient\fP)"
\&
'in -2m
.ZI 2m "\fB-rshift\fP d (\fItranslate\fP)"
\&
'in -2m
'in +2m
\&
.br
These options have affect if neither a row map file nor row
canonification is specified\&. If any of the first two options is used,
indices\ \&\fCi\fP are mapped to\ \&\fCc*i+d\fP\&.
.in -2m

.ZI 2m "\fB-mul\fP e (\fIcoefficient\fP)"
\&
'in -2m
.ZI 2m "\fB-shift\fP f (\fItranslate\fP)"
\&
'in -2m
'in +2m
\&
.br
If a map file is specified for a given domain, neither a map file nor
canonification is specified\&. If any of the first two options is used, the
indices\ \&\fCi\fP will be mapped to\ \&\fCe*i+f\fP\&.
.in -2m

.ZI 2m "\fB-tab\fP fname (\fIread (and map) tab file\fP)"
\&
.br
This option requires the \fB-map\fP option\&. \fBmcxmap\fP will output the
mapped tab definition\&.
.in -2m
.SH EXAMPLES

The matrix below has two canonical domains which are identical\&.
It denotes a map of the canonical domain onto itself, in which
node 0 is relabeled to 8, node 1 is relabeled to 5, et cetera\&.

.nf \fC
(mclheader
mcltype matrix
dimensions 12x12
)
(mclmatrix
begin
0  8  $
1  5  $
2  3  $
3  2  $
4  4  $
5  6  $
6  7  $
7  9  $
8  1  $
9  10 $
10 11 $
11 0  $
)
.fi \fR

.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmcxio(5)\fP,
\fBmcx(1)\fP,
\fBmcxsubs(1)\fP,
and \fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shmx/mcxmap.c
/*   (C) Copyright 2001, 2002, 2003, 2004, 2005 Stijn van Dongen
 *   (C) Copyright 2006, 2007, 2008, 2009, 2010, 2011  Stijn van Dongen
 *   (C) Copyright 2012, 2013 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


#include <string.h>
#include <stdio.h>

#include "impala/matrix.h"
#include "impala/vector.h"
#include "impala/io.h"
#include "impala/iface.h"
#include "impala/app.h"
#include "mcl/interpret.h"

#include "util/io.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/types.h"


enum
{  MY_OPT_IMX
,  MY_OPT_TAB
,  MY_OPT_OUT
,  MY_OPT_MUL
,  MY_OPT_CMUL
,  MY_OPT_RMUL
,  MY_OPT_SHIFT
,  MY_OPT_CSHIFT
,  MY_OPT_RSHIFT
,  MY_OPT_MAP
,  MY_OPT_CMAP
,  MY_OPT_RMAP
,  MY_OPT_MAPI
,  MY_OPT_CMAPI
,  MY_OPT_RMAPI
,  MY_OPT_MAKE_MAP
,  MY_OPT_MAKE_MAPC
,  MY_OPT_MAKE_MAPR
,  MY_OPT_HELP
,  MY_OPT_APROPOS
,  MY_OPT_VERSION
}  ;


const char* me = "mcxmap";
const char* syntax = "Usage: mcxmap [options]";


mcxOptAnchor options[] =
{  {  "-h"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_HELP
   ,  NULL
   ,  "print this help"
   }
,  {  "--version"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_VERSION
   ,  NULL
   ,  "print version information"
   }
,  {  "--help"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_APROPOS
   ,  NULL
   ,  "print this help"
   }
,  {  "-shift"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SHIFT
   ,  "<num>"
   ,  "shift domain indices by <num>"
   }
,  {  "-cshift"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CSHIFT
   ,  "<num>"
   ,  "shift column indices by <num>"
   }
,  {  "-rshift"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RSHIFT
   ,  "<num>"
   ,  "shift row indices by <num>"
   }
,  {  "-mul"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MUL
   ,  "<num>"
   ,  "multiply domain indices by <num>"
   }
,  {  "-cmul"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CMUL
   ,  "<num>"
   ,  "multiply column indices by <num>"
   }
,  {  "-rmul"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RMUL
   ,  "<num>"
   ,  "multiply row indices by <num>"
   }
,  {  "-map"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAP
   ,  "<fname>"
   ,  "map domain indices"
   }
,  {  "-cmap"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CMAP
   ,  "<fname>"
   ,  "map column indices"
   }
,  {  "-rmap"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RMAP
   ,  "<fname>"
   ,  "map row indices"
   }
,  {  "-mapi"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAPI
   ,  "<fname>"
   ,  "map domain indices with inverse"
   }
,  {  "-cmapi"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CMAPI
   ,  "<fname>"
   ,  "map column indices with inverse"
   }
,  {  "-rmapi"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_RMAPI
   ,  "<fname>"
   ,  "map row indices with inverse"
   }
,  {  "-make-map"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAKE_MAP
   ,  "<fname>"
   ,  "canonify domain indices, write map file"
   }
,  {  "-make-mapc"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAKE_MAPC
   ,  "<fname>"
   ,  "canonify column indices, write map file"
   }
,  {  "-make-mapr"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_MAKE_MAPR
   ,  "<fname>"
   ,  "canonify row indices, write map file"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUT
   ,  "<fname>"
   ,  "write output to file <fname>"
   }
,  {  "-imx"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_IMX
   ,  "<fname>"
   ,  "use matrix from file <fname>"
   }
,  {  "-tab"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_TAB
   ,  "<fname>"
   ,  "use tab file from <fname>"
   }
,  {  NULL, 0, 0, NULL, NULL  }
}  ;



int main
(  int                  argc
,  const char*          argv[]
)
   {  mcxIO      *xfin        =  mcxIOnew("-", "r")
   ;  mcxIO      *xfout       =  mcxIOnew("-", "w")
   ;  mclMatrix  *mx          =  NULL
   ;  mclx* cmapx = NULL, *rmapx = NULL
   ;  const char* me          =  "mcxmap"
   ;  long        cshift      =  0
   ;  long        rshift      =  0
   ;  long        cmul        =  1
   ;  long        rmul        =  1
   ;  mcxIO*     xf_cannc     =  NULL
   ;  mcxIO*     xf_cannr     =  NULL
   ;  mcxstatus   status      =  STATUS_OK
   ;  mcxbool     invert      =  FALSE
   ;  mcxbool     invertr     =  FALSE
   ;  mcxbool     invertc     =  FALSE
   ;  mcxIO* xf_map_c = NULL, *xf_map_r = NULL, *xf_map = NULL, *xf_tab = NULL

   ;  mcxOption* opts, *opt
   ;  mcxstatus parseStatus = STATUS_OK

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_IO | MCX_LOG_GAUGE | MCX_LOG_WARN
   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_NO)
   ;  mclx_app_init(stderr)
   
   ;  mcxOptAnchorSortById(options, sizeof(options)/sizeof(mcxOptAnchor) -1)
   ;  opts = mcxOptParse(options, (char**) argv, argc, 1, 0, &parseStatus)

   ;  if (!opts)
      exit(0)

   ;  for (opt=opts;opt->anch;opt++)
      {  mcxOptAnchor* anch = opt->anch

      ;  switch(anch->id)
         {  case MY_OPT_HELP
         :  case MY_OPT_APROPOS
         :  mcxOptApropos(stdout, me, syntax, 0, 0, options)
         ;  return 0
         ;

            case MY_OPT_VERSION
         :  app_report_version(me)
         ;  return 0
         ;

            case MY_OPT_IMX
         :  mcxIOnewName(xfin, opt->val)
         ;  break
         ;

            case MY_OPT_OUT
         :  mcxIOnewName(xfout, opt->val)
         ;  break
         ;

            case MY_OPT_MUL
         :  cmul =  atol(opt->val)
         ;  rmul =  cmul
         ;  break
         ;

            case MY_OPT_CMUL
         :  cmul =  atol(opt->val)
         ;  break
         ;

            case MY_OPT_RMUL
         :  rmul =  atol(opt->val)
         ;  break
         ;

            case MY_OPT_SHIFT
         :  cshift =  atol(opt->val)
         ;  rshift =  atol(opt->val)
         ;  break
         ;

            case MY_OPT_CSHIFT
         :  cshift =  atol(opt->val)
         ;  break
         ;

            case MY_OPT_RSHIFT
         :  rshift =  atol(opt->val)
         ;  break
         ;

            case MY_OPT_MAP
         :  xf_map =  mcxIOnew(opt->val, "r")
         ;  invert =  FALSE
         ;  break
         ;

            case MY_OPT_CMAP
         :  invertc  =  FALSE  
         ;  xf_map_c =  mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_RMAP
         :  invertr  =  FALSE  
         ;  xf_map_r =  mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_MAPI
         :  invert =  TRUE  
         ;  xf_map =  mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_CMAPI
         :  invertc  =  TRUE  
         ;  xf_map_c =  mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_RMAPI
         :  invertr  =  TRUE  
         ;  xf_map_r =  mcxIOnew(opt->val, "r")
         ;  break
         ;

            case MY_OPT_MAKE_MAP
         :  xf_cannc = mcxIOnew(opt->val, "w")
         ;  xf_cannr = xf_cannc
         ;  break
         ;

            case MY_OPT_MAKE_MAPC
         :  xf_cannc = mcxIOnew(opt->val, "w")
         ;  break
         ;

            case MY_OPT_MAKE_MAPR
         :  xf_cannr = mcxIOnew(opt->val, "w")
         ;  break
         ;

            case MY_OPT_TAB
         :  xf_tab = mcxIOnew(opt->val, "r")
         ;  break
         ;
         }
      }

                     /* little special case. restructure when it grows */
      if (xf_tab)
      {  mclTab* tab1, *tab2
      ;  if (xf_map)
         {  mcxIOopen(xf_map, EXIT_ON_FAIL)
         ;  cmapx = mclxRead(xf_map, EXIT_ON_FAIL)  
      ;  }
         else
         mcxDie(1, me, "-tab option requires -map option")

      ;  tab1 = mclTabRead(xf_tab, NULL, EXIT_ON_FAIL)
      ;  if ((tab2 = mclTabMap(tab1, cmapx)))
         mclTabWrite(tab2, xfout, NULL, EXIT_ON_FAIL)
       ; else
         mcxDie(1, me, "map file error (subsumption/bijection)")

      ;  return 0
   ;  }

      mx = mclxRead(xfin, EXIT_ON_FAIL)

   ;  if (xf_map)
      {  mcxIOopen(xf_map, EXIT_ON_FAIL)
      ;  cmapx = mclxRead(xf_map, EXIT_ON_FAIL)  
      ;  rmapx = cmapx
   ;  }
      else
      {  if (xf_map_c)
         {  mcxIOopen(xf_map_c, EXIT_ON_FAIL)
         ;  cmapx = mclxRead(xf_map_c, EXIT_ON_FAIL)  
      ;  }
         else if (cshift || cmul > 1)
         cmapx
         =  mclxMakeMap
            (  mclvCopy(NULL, mx->dom_cols)
            ,  mclvMap(NULL, cmul, cshift, mx->dom_cols)
            )
      ;  else if (xf_cannc)      /* fixme slightly flaky interface */
         {  cmapx 
            =  mclxMakeMap
               (  mclvCopy(NULL, mx->dom_cols)
               ,  mclvCanonical(NULL, mx->dom_cols->n_ivps, 1.0)
               )
         ;  mclxWrite(cmapx, xf_cannc, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL)
      ;  }

         if (xf_map_r)
         {  mcxIOopen(xf_map_r, EXIT_ON_FAIL)
         ;  rmapx = mclxRead(xf_map_r, EXIT_ON_FAIL)  
      ;  }
         else if (rshift || rmul > 1)
         rmapx
         =  mclxMakeMap
            (  mclvCopy(NULL, mx->dom_rows)
            ,  mclvMap(NULL, rmul, rshift, mx->dom_rows)
            )
      ;  else if (xf_cannr)
         {  rmapx 
            =  mclxMakeMap
               (  mclvCopy(NULL, mx->dom_rows)
               ,  mclvCanonical(NULL, mx->dom_rows->n_ivps, 1.0)
               )
         ;  if (xf_cannr != xf_cannc)
            mclxWrite(rmapx, xf_cannr, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL)
         ;  else if (!mclxIsGraph(mx))
            mcxErr(me, "row map not written but matrix is not a graph")
      ;  }
      }

      if (invert && cmapx && cmapx == rmapx)
      {  mclx* cmapxi = mclxTranspose(cmapx)
      ;  mclxFree(&cmapx)
      ;  cmapx = rmapx = cmapxi
   ;  }
      else
      {  if ((invert || invertr) && rmapx)
         {  mclx* rmapxi = mclxTranspose(rmapx)
         ;  mclxFree(&rmapx)
         ;  rmapx = rmapxi
      ;  }
         if ((invert || invertc) && cmapx)
         {  mclx* cmapxi = mclxTranspose(cmapx)
         ;  mclxFree(&cmapx)
         ;  cmapx = cmapxi
      ;  }
      }

   ;  status = STATUS_FAIL

   ;  do
      {  if (cmapx && mclxMapCols(mx, cmapx))
         break
      ;  if (rmapx && mclxMapRows(mx, rmapx))
         break
      ;  status = STATUS_OK
   ;  }
      while (0)

   ;  if (status)
      {  mcxErr(me, "error, nothing written")
      ;  return 1
   ;  }

      mclxWrite(mx, xfout, MCLXIO_VALUE_GETENV, EXIT_ON_FAIL)
   ;  return 0
;  }


